<h2>Why is this an issue?</h2>
<p>This rule allows you to track the usage of <code>noqa</code> comments used to suppress specific linting issues.</p>
<h3>Noncompliant code example</h3>
<pre>
... # noqa: S100
</pre>

