<h2>Why is this an issue?</h2>
<p>Class methods that donâ€™t access instance data can and should be static because they yield more performant code.</p>
<p>To implement a static method in Python one should use either <code>@classmethod</code> or <code>@staticmethod</code>. A class method receives the
class as implicit first argument, just like an instance method receives the instance. A static method does not receive an implicit first argument.</p>
<h3>Noncompliant code example</h3>
<pre>
class Utilities:
    def do_the_thing(self, arg1, arg2, ...):  # Noncompliant
        #...
</pre>
<h3>Compliant solution</h3>
<pre>
class Utilities:
    @classmethod
    def do_the_thing(cls, arg1, arg2, ...):
        #...
</pre>
<p>or</p>
<pre>
class Utilities:
    @staticmethod
    def do_the_thing(arg1, arg2, ...):
        #...
</pre>
<h3>Exceptions</h3>
<p>Methods which raise or may raise a <code>NotImplementedError</code> are ignored.</p>

