<p><code>__all__</code> is used to define the list of module's names that should be imported when <code>from package import *</code> is used. For that
reason, it may only contain strings.</p>
<h2>Noncompliant Code Example</h2>
<pre>
class MyClass:
    pass

__all__ = [
    MyClass  # Noncompliant
]
</pre>
<h2>Compliant Solution</h2>
<pre>
class MyClass:
    pass

__all__ = [
    "MyClass"
]
</pre>
<h2>See</h2>
<ul>
  <li> <a href="https://docs.python.org/3/tutorial/modules.html#importing-from-a-package">Python documentation - Importing * From a Package</a> </li>
</ul>

