<h2>Why is this an issue?</h2>
<p>Operators <a href="https://docs.python.org/3/reference/expressions.html#is-not"><code>is</code></a> and <a
href="https://docs.python.org/3/reference/expressions.html#is-not"><code>is not</code></a> check if their operands point to the same instance, thus
they will always return respectively True and False when they are used to compare objects of different type. Such comparisons can only be bugs.</p>
<h3>Noncompliant code example</h3>
<pre>
myint = 1
mystring = "1"
value = myint is mystring  # Noncompliant. Always False
value = myint is not mystring  # Noncompliant. Always True
</pre>

