<p>This rule raises an issue when an identity comparison operator is used to compare objects of different types.</p>
<h2>Why is this an issue?</h2>
<p>Operators <a href="https://docs.python.org/3/reference/expressions.html#is-not"><code>is</code></a> and <a
href="https://docs.python.org/3/reference/expressions.html#is-not"><code>is not</code></a> check if their operands point to the same instance, thus
they will always return respectively <code>False</code> and <code>True</code> when they are used to compare objects of different types.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
a = 1
b = "1"
value = a is b  # Noncompliant. Always False
value = a is not b  # Noncompliant. Always True
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
a = 1
b = 1
value = a is b
value = a is not b
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>Python documentation - <a href="https://docs.python.org/3/reference/expressions.html#is-not">Identity comparisons</a></li>
</ul>

