<h2>Why is this an issue?</h2>
<p>Having the same implementation in every branch of an <code>if</code> chain is an error. Either a copy-paste error was made and something different
should be executed, or there shouldnâ€™t be an <code>if</code> chain at all.</p>
<h3>Noncompliant code example</h3>
<pre>
if b == 0:  # Noncompliant
    doOneMoreThing()
elif b == 1:
    doOneMoreThing()
else:
    doOneMoreThing()

b = 4 if a &gt; 12 else 4  # Noncompliant
</pre>
<h3>Exceptions</h3>
<p>This rule does not apply to <code>if</code> chains without <code>else</code>-s.</p>
<pre>
if b == 0:  # no issue, this could have been done on purpose to make the code more readable
    doOneMoreThing()
elif b == 1:
    doOneMoreThing()
</pre>

