<h2>Why is this an issue?</h2>
<p>An <code>re.sub</code> call always performs an evaluation of the first argument as a regular expression, even if no regular expression features
were used. This has a significant performance cost and therefore should be used with care.</p>
<p>When <code>re.sub</code> is used, the first argument should be a real regular expression. If it’s not the case, <code>str.replace</code> does
exactly the same thing as <code>re.sub</code> without the performance drawback of the regex.</p>
<p>This rule raises an issue for each <code>re.sub</code> used with a simple string as first argument which doesn’t contains special regex character
or pattern.</p>
<h3>Noncompliant code example</h3>
<pre>
init = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = re.sub(r"Bob is", "It's", init) # Noncompliant
changed = re.sub(r"\.\.\.", ";", changed) # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
init = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = init.replace("Bob is", "It's")
changed = changed.replace("...", ";")
</pre>
<p>Or, with a regex:</p>
<pre>
init = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = re.sub(r"\w*\sis", "It's", init)
changed = re.sub(r"\.{3}", ";", changed)
</pre>

