<h2>Why is this an issue?</h2>
<p>Calling an operator in python is equivalent to calling a special method (except for the identity operator <code>is</code>). Python provides a set
of built-in operations. It is for example possible to add two integers: <code>1 + 2</code>. It is however not possible to add a string and an integer:
<code>1 + "2"</code> and such an operation will raise a <code>TypeError</code>.</p>
<p>It is possible to define how an operator will behave with a custom class by defining the corresponding special method. See python documentation for
a complete list of operators and their methods: <a
href="https://docs.python.org/3/reference/datamodel.html?highlight=__exit__%20special%20methods#emulating-numeric-types">arithmetic and bitwise
operators</a>, <a href="https://docs.python.org/3/reference/datamodel.html?highlight=__exit__%20special%20methods#object.__lt__">comparison
operators</a>.</p>
<p>For symmetrical binary operators you need to define two methods so that the order of operands doesnâ€™t matter, ex: <code>__add__</code> and
<code>__radd__</code>.</p>
<p>This rule raises an issue when an operator is used on incompatible types. Types are considered incompatible if no built-in operations between those
types exist and none of the operands has implemented the corresponding special methods.</p>
<h3>Noncompliant code example</h3>
<pre>
class Empty:
    pass

class Add:
    def __add__(self, other):
        return 42

1 + 2
1 + "2"  # Noncompliant
Empty() + 1  # Noncompliant
Add() + 1
1 + Add()  # Noncompliant
Add() + Empty()
Empty() + Add()  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
class Empty:
    pass

class Add:
    def __add__(self, other):
        return 42

    def __radd__(self, other):
        return 42

Add() + 1
1 + Add()
Add() + Empty()
Empty() + Add()
</pre>
<h2>Resources</h2>
<ul>
  <li> Python documentation - <a href="https://docs.python.org/3/reference/datamodel.html?highlight=__exit__%20special%20methods#object.__lt__">Rich
  comparison methods</a> </li>
  <li> Python documentation - <a
  href="https://docs.python.org/3/reference/datamodel.html?highlight=__exit__%20special%20methods#emulating-numeric-types">Emulating numeric types</a>
  </li>
</ul>

