<p>This rule raises an issue when an instance or a class method does not have at least one positional parameter.</p>
<h2>Why is this an issue?</h2>
<p>Every instance method is expected to have at least one positional parameter. This parameter will reference the object instance on which the method
is called. Calling an instance method which doesnâ€™t have at least one parameter will raise a <code>TypeError</code>. By convention, this first
parameter is usually named <code>self</code>.</p>
<p>Class methods, i.e. methods annotated with <code>@classmethod</code>, also require at least one parameter. The only differences is that they will
receive the class itself instead of a class instance. By convention, this first parameter is usually named <code>cls</code>.</p>
<h3>Exceptions</h3>
<p>Static methods (methods annotated with the <code>@staticmethod</code> decorator) do not require any positional parameter. This rule will not raise
an issue on them.</p>
<h2>How to fix it</h2>
<p>Adding <code>self</code> as the first parameter of an instance method or <code>cls</code> as the first parameter of a class method will resolve the
issue.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
class MyClass:
    def instance_method():  # Noncompliant: "self" parameter is missing.
        print("instance_method")

    @classmethod
    def class_method():  # Noncompliant: "cls" parameter is missing.
        print("class_method")
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
class MyClass:
    def instance_method(self):
        print("instance_method")

    @classmethod
    def class_method(cls):
        print("class_method")

    @staticmethod
    def static_method():
        print("static_method")
</pre>
<h3>Pitfalls</h3>
<p>Note that <code>__new__</code> and <code>__init_subclass__</code> take a class as first argument even thought they are not decorated with
<code>@classmethod</code>.</p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li><a href="https://docs.python.org/3.11/tutorial/classes.html#method-objects">Method Objects</a> - Python Method Objects</li>
  <li><a href="https://docs.python.org/3.11/library/functions.html?highlight=classmethod#classmethod">Class Method</a> - Python Class Method</li>
</ul>

