<h2>Why is this an issue?</h2>
<p>Property getter, setter and deleter methods are called by the python interpreter with a specific number or arguments:</p>
<ul>
  <li> only "self" for property getter and deleter methods. </li>
  <li> "self" and a value for setter methods. </li>
</ul>
<p>Adding any other parameter, or removing these mandatory parameters will make method calls fail.</p>
<p>This rule raises an issue when:</p>
<ul>
  <li> too many parameters are defined in a property getter, setter or deleter method. </li>
  <li> the value parameter is missing in a property setter method. </li>
</ul>
<h3>Noncompliant code example</h3>
<pre>
class A:
    @property
    def foo(self, unexpected, unexpected2):  # Noncompliant. Too many parameters.
        return self._foo

    @foo.setter
    def foo(self, value, unexpected):  # Noncompliant. Too many parameters.
        self._foo = value

    @foo.deleter
    def foo(self, unexpected):  # Noncompliant. Too many parameters.
        del self._foo

class B:
    def get_foo(self, unexpected):  # Noncompliant. Too many parameters.
        return self._foo

    def set_foo(self, value, unexpected):  # Noncompliant. Too many parameters.
        self._foo = value

    def del_foo(self, unexpected):  # Noncompliant. Too many parameters.
        del self._foo

    foo = property(get_foo, set_foo, del_foo, "'foo' property.")
</pre>
<h3>Compliant solution</h3>
<pre>
class A:
    @property
    def foo(self):
        return self._foo

    @foo.setter
    def foo(self, value):
        self._foo = value

    @foo.deleter
    def foo(self):
        del self._foo

class B:
    def get_foo(self):
        return self._foo

    def set_foo(self, value):
        self._foo = value

    def del_foo(self):
        del self._foo

    foo = property(get_foo, set_foo, del_foo, "'foo' property.")
</pre>
<h2>Resources</h2>
<ul>
  <li> <a href="https://docs.python.org/3/library/functions.html#property">Python Documentation - Built-in Functions - property</a> </li>
</ul>

