<p>This rule raises an issue when a bare <code>raise</code> statement is not in an <code>except</code> or <code>finally</code> block.</p>
<h2>Why is this an issue?</h2>
<p>A bare <code>raise</code> statement, i.e. a <code>raise</code> with no exception provided, will re-raise the last active exception in the current
scope:</p>
<pre>
def foo():
    try:
        ...
    except ValueError as e:
        raise # this will re-raise "e"
</pre>
<p>If the <code>raise</code> statement is not in an <code>except</code> or <code>finally</code> block, no exception is active and a
<code>RuntimeError</code> is raised instead.</p>
<p>If the bare <code>raise</code> statement is in a function called in an <code>except</code> block, the exception caught by the <code>except</code>
will be re-raised. However, this behavior is not reliable as nothing prevents a developer from calling the function from a different context.</p>
<p>Overall, having bare <code>raise</code> statements outside of <code>except</code> blocks is discouraged as it is hard to understand and
maintain.</p>
<h3>Notes</h3>
<p>In a <code>finally</code> block, an exception is still active only when it hasnâ€™t been caught in a previous <code>except</code> clause or if it has
been raised in an <code>except</code> block. In both cases, it is better to let the exception propagate automatically than to re-raise it. This
pattern is covered by rule {rule:python:S5704}.</p>
<h2>How to fix it</h2>
<p>To fix this issue, make sure to specify which exception needs to be raised when outside of an <code>except</code> block.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
raise  # Noncompliant: no active exception

def foo():
    raise  # Noncompliant: no active exception
    try:
        raise  # Noncompliant: no active exception
    except ValueError:
        handle_error()

def handle_error():
    raise  # Noncompliant: this is not reliable
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
raise ValueError()

def foo():
    raise ValueError()
    try:
        raise ValueError()
    except ValueError:
        raise
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>Python Documentation - <a href="https://docs.python.org/3/reference/simple_stmts.html#raise">The <code>raise</code> statement</a></li>
</ul>

