<h2>Why is this an issue?</h2>
<p>Identity operators <code>is</code> and <code>is not</code> check if the same object is on both sides, i.e. <code>a is b</code> returns
<code>True</code> if <code>id(a) == id(b)</code>.</p>
<p>Integers, bytes, floats, strings, frozensets and tuples should not be compared with identity operators because the result may not be as expected.
If you need to compare these types you should use instead equality operators <code>==</code> or <code>!=</code>.</p>
<p>The CPython interpreter caches certain builtin values for integers, bytes, floats, strings, frozensets and tuples. For example, the literal
<code>1</code> will create the same object as <code>int("1")</code>, which means that <code>1 is int("1")</code> is True. However this works only by
chance as other integer values are not cached, for example <code>int("1000") is 1000</code> will always be <code>False</code>. This behavior is not
part of Python language specification and could vary between interpreters. CPython 3.8 even <a
href="https://docs.python.org/3.8/whatsnew/3.8.html#changes-in-python-behavior">warns about comparing literals using identity operators</a>.</p>
<p>The only case where using the "is" operator with a cached type is ok is with "interned" strings. Note however that interned strings donâ€™t
necessarily have the same identity as string literals.</p>
<p>This rule raises an issue when at least one operand of an identity operator:</p>
<ul>
  <li> is of type <code>int</code>, <code>bytes</code>, <code>float</code>, <code>frozenset</code> or <code>tuple</code>. </li>
  <li> or it is a string literal. </li>
</ul>
<h3>Noncompliant code example</h3>
<pre>
def literal_comparison(param):
    param is 2000  # Noncompliant

literal_comparison(2000)  # will return True
literal_comparison(int("2000"))  # will return False

() is tuple()  # Noncompliant. Always True
(1,) is tuple([1])  # Noncompliant. Always False

from sys import intern

with open("test.txt") as f:  # test.txt contains "blabla\n"
    text = f.read()
intern(text) is "blabla\n"  # Noncompliant. Always False
</pre>
<h3>Compliant solution</h3>
<pre>
def literal_comparison(param):
    param == 2000

literal_comparison(2000)  # will return True
literal_comparison(int("2000"))  # will return True

() == tuple()  # Always True
(1,) == tuple([1])  # Always True

from sys import intern

with open("tmp/test.txt") as f:  # test.txt contains "blabla\n"
    text = f.read()
intern(text) is intern("blabla\n")  # Always True
</pre>
<h2>Resources</h2>
<ul>
  <li> <a href="https://adamj.eu/tech/2020/01/21/why-does-python-3-8-syntaxwarning-for-is-literal/">Why does Python 3.8 log a SyntaxWarning for 'is'
  with literals?</a> - Adam Johnson </li>
  <li> <a href="https://treyhunner.com/2019/03/unique-and-sentinel-values-in-python/#Equality_vs_identity">Equality vs identity</a> - Trey Hunner
  </li>
  <li> <a href="https://docs.python.org/3.7/library/sys.html?highlight=sys.intern#sys.intern">Python documentation - sys.intern</a> </li>
</ul>

