<h2>Why is this an issue?</h2>
<p>Python concatenates adjacent string or byte literals at compile time. It means that <code>"a" "b"</code> is equivalent to <code>"ab"</code>. This
is sometimes used to split a long string on multiple lines. However an implicit string concatenation can also be very confusing. In the following
contexts it might indicate that a comma was forgotten:</p>
<ul>
  <li> when the two strings are on the same line it looks like a badly formatted tuple. Parenthesises are not mandatory to create a tuple, only the
  comma is. </li>
  <li> when the strings are in a list, set or tuple. </li>
</ul>
<h3>Noncompliant code example</h3>
<pre>
def func():
    return "item1" "item2"  # Noncompliant

["1",
 "2"  # Noncompliant
 "3",
 "a very very very"  # Noncompliant
 "very very long string",
 "4"]
</pre>
<h3>Compliant solution</h3>
<pre>
def func():
    return "item1", "item2"

["1",
 "2",
 "3",
 "a very very very" +
 "very very long string",
 "4"]
</pre>
<h3>Exceptions</h3>
<p>No issue will be raised when there is a visible reason for the string concatenation:</p>
<ul>
  <li> when the quotes used for both strings are different. This can be used to avoid escaping quotes </li>
  <li> when the strings or bytes have different prefixes, i.e. "f" for f-strings, "r" for raw, "u" for unicode and no prefix for normal strings. </li>
  <li> when strings are visibly split to avoid long lines of code. (Example: the first string ends with a space, punctuation or <code>\n</code>).
  </li>
</ul>

