<p>This rule raises an issue when a constant expression is used as a condition in an <code>if</code>, <code>elif</code>, a conditional expression or
other boolean expressions.</p>
<h2>Why is this an issue?</h2>
<p>When a constant is used as a condition, either it has no effect on the execution flow and it can be removed, or some code will never be executed
and it is a bug.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def foo():
    a = True
    if a: # Noncompliant: the condition is always true
      return 1
    else:
      return 2
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def foo():
    a = bar()
    if a:
      return 1
    else:
      return 2
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>Python documentation - <a href="https://www.python.org/dev/peps/pep-0285/">PEP 285 - Adding a bool type</a></li>
  <li>Python documentation - <a href="https://docs.python.org/3/library/stdtypes.html#truth-value-testing">Python documentation - Truth Value
  Testing</a></li>
</ul>

