<p>This rule raises an issue when a name listed in the <code>__all__</code> property of a module has not been defined.</p>
<h2>Why is this an issue?</h2>
<p>The <code>__all__</code> property of a module is used to define the list of names that will be imported when performing a wildcard import of this
module, i.e. when <code>from mymodule import *</code> is used.</p>
<p>In the following example:</p>
<pre>
# mymodule.py

def foo(): ...
def bar(): ...

__all__ = ["foo"]
</pre>
<p>Executing <code>from mymodule import *</code> from a different module will only import <code>foo</code>.</p>
<p>This list can only reference defined names, otherwise an <code>AttributeError</code> will be raised when the module is imported.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre>
from mymodule import my_func

__all__ = ["unknown_func"]  # Noncompliant: "unknown_func" is undefined
</pre>
<h4>Compliant solution</h4>
<pre>
from mymodule import my_func

__all__ = ["my_func"]
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li><a href="https://docs.python.org/3/tutorial/modules.html#importing-from-a-package">Python documentation - importing * from a package</a></li>
</ul>

