<p>This rule raises an issue when an assigned value type is not compatible with the type hint of the assigning variable.</p>
<h2>Why is this an issue?</h2>
<p>Type hints in Python allow you to specify the expected types of variables and function return values. While type hints are not enforced at runtime,
they serve as documentation and can be checked using static type checkers to catch type-related errors during development.</p>
<p>When an assigned value type is incompatible with the type hint of the assigning variable, it can lead to several issues:</p>
<ul>
  <li><strong>Type-related bugs:</strong> Assigning a value of an incompatible type to a variable with a specific type hint may lead to unexpected
  behavior or errors at runtime.</li>
  <li><strong>Readability and maintainability:</strong> Type hints improve code readability by explicitly stating the intended types of variables and
  functions. When the assigned value type doesn’t match the hint, it can confuse other developers and make the code harder to maintain.</li>
</ul>
<h2>How to fix it</h2>
<p>Assign the variable to the value compatible with the type hint or change the type hint to be compatible with the variable’s type.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def my_function():
    my_int: int = "string"  # Noncompliant
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def my_function():
    my_str: str = "string"
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>Python documentation - <a href="https://docs.python.org/3/library/typing.html">typing — Support for type hints</a></li>
</ul>

