<p>This rule enforces that the <code>@receiver</code> decorator in Django and the <code>@route</code> decorator in Flask are placed on top of all
other decorators.</p>
<h2>Why is this an issue?</h2>
<p>The order in which decorators are applied can have a significant impact on their behavior. When decorators are applied in Python, they are
processed from bottom to top. The outermost decorator (the one written first) is applied last and receives the result of all inner decorators.</p>
<p>In Django, the <code>@receiver</code> decorator is used to register signal handlers. It must be the outermost decorator (written at the top) so
that all other decorators are applied to the function before it gets registered as a signal handler. If <code>@receiver</code> is not outermost, it
registers the original function, and the decorators written above it will not affect the signal handler, resulting in unexpected behavior or
errors.</p>
<p>In Flask, the <code>@app.route()</code> or <code>@blueprint.route()</code> decorator must be the outermost decorator for view functions. If other
decorators are placed outside the route decorator, the route may not be registered correctly, leading to 404 errors or routes that donâ€™t respond to
requests as expected.</p>
<h2>How to fix it in Django</h2>
<p>Move the <code>@receiver</code> decorator to the top of the list of decorators used to decorate the function.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
from django.dispatch import receiver
from django.views.decorators.csrf import csrf_exempt

@csrf_exempt
@receiver(some_signal)  # Noncompliant: @receiver should be the outermost decorator
def my_handler(sender, **kwargs):
    ...
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
from django.dispatch import receiver
from django.views.decorators.csrf import csrf_exempt

@receiver(some_signal)
@csrf_exempt
def my_handler(sender, **kwargs):
    ...
</pre>
<h2>How to fix it in Flask</h2>
<p>Move the <code>@app.route()</code> or <code>@blueprint.route()</code> decorator to be the outermost (first) decorator.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
@login_required
@app.route('/secret_page')  # Noncompliant: @route should be the outermost decorator
def secret_page():
    return "Secret content"
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
@app.route('/secret_page')
@login_required
def secret_page():
    return "Secret content"
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Django - <a href="https://docs.djangoproject.com/en/4.1/topics/signals/">Signals</a> </li>
  <li> Flask - <a href="https://flask.palletsprojects.com/en/stable/patterns/viewdecorators/">View Decorators</a> </li>
  <li> Flask - <a href="https://flask.palletsprojects.com/en/stable/quickstart/#routing">Routing</a> </li>
</ul>

