<p>This rule raises an issue when an object used as a sequence index does not define an <code>__index__</code> method.</p>
<h2>Why is this an issue?</h2>
<p>Objects can be used as sequence indexes to access a specific element from the sequence, through the following syntax:</p>
<pre>
my_list = [1, 2, 3, 4]
x = 1
print(my_list[x])  # This will print 2
</pre>
<p>Whenever an object is used as a sequence index, the Python interpreter calls its <code>__index__</code> method to compute the index that needs to
be accessed from the sequence.</p>
<p>Any object can be used as sequence index, as long as it defines an <code>__index__</code> method that returns an <code>int</code>. Most commonly,
sequence indexes are simply integers.</p>
<p>Similarly, sequences can be sliced through the following syntax:</p>
<pre>
my_list = [1, 2, 3, 4]
x = 1
print(my_list[1:3]) # This will print [2, 3]
</pre>
<p>If an invalid object is used as a sequence index, a <code>TypeError</code> will be raised.</p>
<h2>How to fix it</h2>
<p>Make sure to use an object that defines an <code>__index__</code> method as sequence indexes.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def foo():
    my_list = ["spam", "eggs"]
    x = my_list["spam"]  # Noncompliant: a str is not a valid index
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def foo():
    my_list = ["spam", "eggs"]
    x = my_list[0]
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<p>Python Documentation - <a href="https://docs.python.org/3/library/operator.html#operator.<em>index</em>"><em>index</em> method</a></p>

