<p>This rule raises an issue when <code>np.where</code> is used with only the condition parameter set.</p>
<h2>Why is this an issue?</h2>
<p>The NumPy function <code>np.where</code> provides a way to execute operations on an array under a certain condition:</p>
<pre>
import numpy as np

arr = np.array([1,2,3,4])

result = np.where(arr &gt; 3, arr * 2, arr)
</pre>
<p>In the example above the <code>np.where</code> function will multiply all the elements in the array which satisfy the condition: <code>element &gt;
3</code> by 2. The elements that do not satisfy the condition will be left untouched. The <code>result</code> array holds now the values 1, 2, 3 and
8.</p>
<p>It is also possible to call <code>np.where</code> with only the condition parameter set:</p>
<pre>
import numpy as np

arr = np.array([1,2,3,4])

result = np.where(arr &gt; 2)
</pre>
<p>Even though this is perfectly valid code in NumPy, it may not yield the expected results.</p>
<p>When providing only the condition parameter to the <code>np.where</code> function, it will behave as <code>np.asarray(condition).nonzero()</code>
or <code>np.nonzero(condition)</code>. Both these functions provide a way to find the indices of the elements satisfying the condition passed as
parameter. Be mindful that <code>np.asarray(condition).nonzero()</code> and <code>np.nonzero(condition)</code> do not return the
<strong>values</strong> that satisfy the condition but only their <strong>indices</strong>. This means the <code>result</code> variable now holds a
tuple with the first element being an array of all the indices where the condition <code>arr &gt; 2</code> was satisfied:
<code>(array([2,3]),)</code>.</p>
<p>If the intention is to find the indices of the elements which satisfy a certain condition it is preferable to use the
<code>np.asarray(condition).nonzero()</code> or <code>np.nonzero(condition)</code> function instead.</p>
<h2>How to fix it</h2>
<p>To fix this issue either:</p>
<ul>
  <li>provide all three parameters to the <code>np.where</code> function (condition, value if the condition is satisfied, value if the condition is
  not satisfied) or,</li>
  <li>use the <code>np.nonzero</code> function.</li>
</ul>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import numpy as np

def bigger_than_two():
  arr = np.array([1,2,3,4])
  result = np.where(arr &gt; 2) # Noncompliant: only the condition parameter is provided to the np.where function.
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import numpy as np

def bigger_than_two():
  arr = np.array([1,2,3,4])
  result = np.where(arr &gt; 2, arr + 1, arr) # Compliant
  indices = np.nonzero(arr &gt; 2) # Compliant
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>NumPy Documentation - <a href="https://numpy.org/doc/stable/reference/generated/numpy.where.html#numpy-where">numpy.where</a></li>
  <li>NumPy Documentation - <a href="https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html#numpy.nonzero">numpy.nonzero</a></li>
</ul>

