<p>This rule raises an issue when passing a <code>pytz.timezone</code> to the <code>datetime.datetime</code> constructor.</p>
<h2>Why is this an issue?</h2>
<p>When working with timezones in Python, it’s important to understand that the <code>datetime.datetime</code> constructor and <code>pytz</code>
timezone objects handle timezones differently. This difference can lead to unexpected results if a <code>pytz</code> object is used as the
<code>tzinfo</code> argument in the <code>datetime.datetime</code> constructor.</p>
<p>The <code>datetime.datetime</code> constructor expects a <code>tzinfo</code> object that is a subclass of the <code>datetime.tzinfo</code> base
class. <code>pytz</code> timezone objects do provide this interface, but they implement it in a way that’s not fully compatible with
<code>datetime.datetime</code>.</p>
<p>One key difference is how they handle historical changes in timezone offsets. The <code>datetime</code> module uses the IANA time zone database,
which includes historical changes.</p>
<p>When you create a <code>datetime</code> object with a <code>pytz</code> timezone object as the <code>tzinfo</code> argument, it uses the earliest
known offset for that timezone. This can lead to unexpected offsets, as the earliest known offset may not match the current or most commonly used
offset for that timezone.</p>
<p>For example, if you were to use 'US/Eastern' as your timezone, you might expect the offset to be either -5 hours (Eastern Standard Time) or -4
hours (Eastern Daylight Time), depending on the time of the year. However, due to historical changes, the actual offset might be something different,
like -4 hours and 56 minutes. This can lead to subtle bugs in your code, especially if you’re doing calculations with datetime objects.</p>
<p>Note that, when using Python 3.9 and later, it is recommended to use the <code>zoneinfo</code> package from the standard library over
<code>pytz</code>.</p>
<h2>How to fix it</h2>
<p>To avoid these issues, it’s recommended to use the localize method of <code>pytz</code> timezone objects to attach a timezone to a
<code>datetime</code> object. This method correctly handles historical changes and other timezone-related issues.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import datetime
import pytz

dt = datetime.datetime(2022, 1, 1, tzinfo=pytz.timezone('US/Eastern'))  # Noncompliant: 2022-01-01 00:00:00-04:56
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import datetime
import pytz

dt = pytz.timezone('US/Eastern').localize(datetime.datetime(2022, 1, 1))  # OK: 2022-01-01 00:00:00-05:00
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>PEP 615 - <a href="https://peps.python.org/pep-0615/">Support for the IANA Time Zone Database in the Standard Library</a></li>
</ul>

