<p>This rule raises an issue when synchronous HTTP client calls are used within asynchronous functions.</p>
<h2>Why is this an issue?</h2>
<p>Using synchronous HTTP clients like <code>urllib3</code>, <code>requests</code>, or <code>httpx.Client</code> in asynchronous code blocks the
entire event loop. This undermines the primary advantage of asynchronous programming - the ability to perform concurrent operations without blocking
execution.</p>
<p>When an async function makes a synchronous HTTP request:</p>
<ul>
  <li>The event loop is completely blocked until the HTTP operation completes</li>
  <li>No other coroutines can run during this time, even if theyâ€™re ready to execute</li>
  <li>The responsiveness of the application is degraded</li>
  <li>In server applications, this can cause timeouts or failures for other concurrent requests</li>
</ul>
<p>Instead, async-compatible HTTP clients should be used:</p>
<ul>
  <li><code>httpx.AsyncClient</code> - works with asyncio, Trio, and AnyIO</li>
  <li><code>aiohttp.ClientSession</code> - works with asyncio</li>
  <li><code>asks</code> - works with Trio and asyncio</li>
</ul>
<p>Using these libraries allows other tasks to continue executing while waiting for HTTP responses, significantly improving application performance
and responsiveness.</p>
<h2>How to fix it</h2>
<p>Replace synchronous HTTP clients with asynchronous alternatives. The <code>httpx.AsyncClient</code> is recommended as it provides a consistent API
across asyncio, Trio, and AnyIO frameworks.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import requests

async def fetch_data():
    response = requests.get("https://api.example.com/data")  # Noncompliant
    return response.json()
</pre>
<h4>Compliant solution</h4>
<p>Using httpx.AsyncClient (works with asyncio, Trio, and AnyIO):</p>
<pre data-diff-id="1" data-diff-type="compliant">
import httpx

async def fetch_data():
    async with httpx.AsyncClient() as client:
        response = await client.get("https://api.example.com/data")
        return response.json()
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li>HTTPX - <a href="https://www.python-httpx.org/async/">Async Usage</a></li>
  <li>AIOHTTP - <a href="https://docs.aiohttp.org/en/stable/client_quickstart.html">Client Quickstart</a></li>
  <li>Asks - <a href="https://asks.readthedocs.io/en/latest/">Documentation</a></li>
</ul>
<h3>Articles &amp; blog posts</h3>
<ul>
  <li>Python - <a href="https://realpython.com/async-io-python/">Async IO in Python</a></li>
</ul>

