<p>This rule raises an issue when a cancellation scope (timeout context) is used without any checkpoints making the timeout functionality
ineffective.</p>
<h2>Why is this an issue?</h2>
<p>When using asynchronous programming with libraries like <code>trio</code> or <code>anyio</code>, cancellation scopes (timeout contexts) are used to
implement timeouts and cancellation. However, these mechanisms only work when there are checkpoints within the scope where cancellation can occur.
Without any checkpoints, the timeout will never be triggered, making it ineffective.</p>
<p>A checkpoint is a point in the code where cancellation can be detected and acted upon. Common checkpoints include:</p>
<ul>
  <li> Explicit calls to the checkpoint method of the used framework </li>
  <li> <code>yield</code> statements </li>
  <li> <code>await</code> statements </li>
</ul>
<h3>What is the potential impact?</h3>
<p>Without proper checkpoints in cancel scopes:</p>
<ul>
  <li> Timeouts won’t work as expected </li>
  <li> Resources might be held longer than intended </li>
  <li> The application might become unresponsive or hang </li>
  <li> Cancellation operations won’t be honored </li>
</ul>
<h2>How to fix it in Asyncio</h2>
<p>There is no direct checkpoint method in <code>asyncio</code>, but you can use <code>await asyncio.sleep(0)</code> as a workaround.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="3" data-diff-type="noncompliant">
import asyncio

async def process_data(data):
    try:
        async with asyncio.timeout(1.0):  # Noncompliant
            result = expensive_computation(data)
            return result
    except asyncio.TimeoutError:
        return None
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="3" data-diff-type="compliant">
import asyncio

async def process_data(data):
    try:
        async with asyncio.timeout(1.0):  # Compliant
            result = expensive_computation(data)
            await asyncio.sleep(0)
            return result
    except asyncio.TimeoutError:
        return None
</pre>
<h2>How to fix it in Trio</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import trio

async def process_data(data):
    async with trio.move_on_after(1.0):  # Noncompliant
        result = expensive_computation(data)
        return result
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import trio

async def process_data(data):
    async with trio.move_on_after(1.0):  # Compliant
        result = expensive_computation(data)
        await trio.lowlevel.checkpoint()
        return result
</pre>
<h2>How to fix it in AnyIO</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
import anyio

async def process_data(data):
    async with anyio.move_on_after(1.0):  # Noncompliant
        result = expensive_computation(data)
        return result
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
import anyio

async def process_data(data):
    async with anyio.move_on_after(1.0):  # Compliant
        result = expensive_computation(data)
        await anyio.lowlevel.checkpoint()
        return result
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Trio documentation - <a href="https://trio.readthedocs.io/en/stable/reference-core.html#cancellation-and-timeouts">Cancellation and
  timeouts</a> </li>
  <li> AnyIO documentation - <a href="https://anyio.readthedocs.io/en/stable/cancellation.html#timeouts">Timeouts</a> </li>
</ul>

