<p>This rule raises an issue when synchronous file operations like <code>open()</code> are used within asynchronous functions.</p>
<h2>Why is this an issue?</h2>
<p>Using synchronous file operations like <code>open()</code> in asynchronous code blocks the entire event loop. This undermines the primary advantage
of asynchronous programming - the ability to perform concurrent operations without blocking execution.</p>
<p>When an async function makes a synchronous file operation:</p>
<ul>
  <li> The event loop is completely blocked until the file I/O operation completes </li>
  <li> No other coroutines can run during this time, even if they’re ready to execute </li>
  <li> The responsiveness of the application is degraded </li>
  <li> In server applications, this can cause timeouts or failures for other concurrent requests </li>
</ul>
<p>Instead, async libraries provide mechanisms to handle file operations asynchronously:</p>
<ul>
  <li> <code>aiofiles</code> library for asyncio </li>
  <li> <code>trio.open_file()</code> for Trio </li>
  <li> <code>anyio.open_file()</code> for AnyIO </li>
</ul>
<p>Using these constructs allows other tasks to continue executing while waiting for the potentially blocking file operation to complete.</p>
<h2>How to fix it in Asyncio</h2>
<p>The standard library’s <code>asyncio</code> package doesn’t provide built-in asynchronous file I/O operations. Use the <code>aiofiles</code>
library to handle file operations asynchronously.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
async def read_config():
    with open("config.json", "r") as file:  # Noncompliant
        data = file.read()
    return data
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import aiofiles

async def read_config():
    async with aiofiles.open("config.json", "r") as file: # Compliant
        data = await file.read()
    return data
</pre>
<h2>How to fix it in Trio</h2>
<p>Use <code>trio.open_file()</code> to handle file operations asynchronously.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
async def read_config():
    with open("config.json", "r") as file:  # Noncompliant
        data = file.read()
    return data
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
import trio

async def read_config():
    async with await trio.open_file("config.json", "r") as file: # Compliant
        data = await file.read()
    return data
</pre>
<h2>How to fix it in AnyIO</h2>
<p>Use <code>anyio.open_file()</code> to handle file operations asynchronously.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="3" data-diff-type="noncompliant">
async def read_config():
    with open("config.json", "r") as file:  # Noncompliant
        data = file.read()
    return data
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="3" data-diff-type="compliant">
import anyio

async def read_config():
    async with await anyio.open_file("config.json", "r") as file: # Compliant
        data = await file.read()
    return data
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Trio - <a href="https://trio.readthedocs.io/en/stable/reference-io.html#trio.open_file">trio.open_file() documentation</a> </li>
  <li> AnyIO - <a href="https://anyio.readthedocs.io/en/stable/fileio.html">File I/O in AnyIO</a> </li>
  <li> Aiofiles - <a href="https://github.com/Tinche/aiofiles">aiofiles project</a> </li>
</ul>
<h3>Articles &amp; blog posts</h3>
<ul>
  <li> Python - <a href="https://realpython.com/async-io-python/">Async IO in Python</a> </li>
</ul>

