<p>This rule raises an issue when a function is declared async but does not use any asynchronous features.</p>
<h2>Why is this an issue?</h2>
<p>Declaring a function with <code>async def</code> signals that it’s a coroutine, typically because it needs to use <code>await</code> for
non-blocking operations (like I/O), or employs other asynchronous features like <code>async for</code> or <code>async with</code>.</p>
<p>If none of these asynchronous mechanisms are utilized within the function’s body, the <code>async</code> declaration is superfluous.</p>
<p>Using <code>async</code> unnecessarily can:</p>
<ul>
  <li> <strong>Reduce Code Clarity:</strong> It misleads developers into thinking the function performs asynchronous operations when it doesn’t,
  making the codebase harder to understand and reason about. </li>
  <li> <strong>Introduce Minor Overhead:</strong> Calling an <code>async def</code> function creates a coroutine object. While modern Python is
  efficient, this still incurs more overhead than a direct synchronous function call if the asynchronicity isn’t actually leveraged. The function’s
  body will execute synchronously until an <code>await</code> (if any) is encountered. If there’s no <code>await</code> or other async feature, it
  essentially runs synchronously but wrapped as a coroutine. </li>
  <li> <strong>Complicate Usage:</strong> Callers must use <code>await</code> (or schedule it as a task) to execute the coroutine and get its result.
  This is an unnecessary ceremony if the function itself isn’t truly asynchronous. </li>
</ul>
<h2>How to fix it</h2>
<p>Either remove the <code>async</code> keyword, or start using the appropriate asynchronous features.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
async def my_function():  # Noncompliant
    print("Hello from my function")
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def my_function():  # Compliant
    print("Hello from my function")
</pre>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
def another_function(): ...

async def my_function():  # Noncompliant
    return another_function()
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
async def another_function(): ...

async def my_function():  # Compliant
    return await another_function()
</pre>
<h2>Resources</h2>
<h3>Articles &amp; blog posts</h3>
<ul>
  <li> <a href="https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/">What Color Is Your Function? - Bob Nystrom</a> </li>
</ul>

