<p>This rule raises an issue when an iterable is reversed using slicing, like <code>[::-1]</code>, before being passed to <code>set()</code>,
<code>sorted()</code>, or <code>reversed()</code>.</p>
<h2>Why is this an issue?</h2>
<p>Reversing an iterable using slicing, likle <code>[::-1]</code>, before passing it to <code>set()</code>, <code>sorted()</code>, or
<code>reversed()</code> is unnecessary and inefficient. The slicing operation creates a new copy of the iterable in reverse order, which is not needed
for the following reasons:</p>
<ul>
  <li> <code>set()</code>: The order of elements in a set is inherently undefined, so reversing the iterable before creating the set has no effect on
  the final set, and it introduces unnecessary computation. </li>
  <li> <code>sorted()</code>: The <code>sorted()</code> function has a <code>reverse</code> parameter that provides a more efficient way to sort in
  descending order. Using slicing to reverse the result of <code>sorted()</code> is less efficient and less readable. </li>
  <li> <code>reversed()</code>: Applying <code>reversed()</code> twice on the same iterable effectively returns the original iterable, if it supports
  direct iteration. If the iterable is a one-time iterator, then you will need to create an iterator from the original iterable using
  <code>iter(iterable)</code>. Using slicing adds unnecessary overhead. </li>
</ul>
<h2>How to fix it</h2>
<p>To fix these issues remove the redundant slicing operation: * <code>set(iterable[::-1])</code>: Replace with <code>set(iterable)</code>. *
<code>sorted(iterable)[::-1]</code>: Replace with <code>sorted(iterable, reverse=True)</code>. * <code>reversed(iterable[::-1])</code>: Replace with
<code>iterable</code> if it supports direct iteration, or <code>iter(iterable)</code> if it is a one-time iterator.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
iterable = [1, 3, 2]
result = set(iterable[::-1]) # Noncompliant
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
iterable = [1, 3, 2]
result = set(iterable)
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Python Documentation - <a href="https://docs.python.org/3/library/functions.html#sorted">sorted</a> </li>
  <li> Python Documentation - <a href="https://docs.python.org/3/library/functions.html#reversed">reversed</a> </li>
  <li> Python Documentation - <a href="https://docs.python.org/3/library/functions.html#func-set">set</a> </li>
</ul>

