<p>This rule raises when a control flow statement (<code>return</code>, <code>break</code>, <code>continue</code>) is used inside a TaskGroup or
Nursery context manager.</p>
<h2>Why is this an issue?</h2>
<p>Using control flow statements like <code>return</code>, <code>break</code>, or <code>continue</code> inside async TaskGroup or Nursery blocks leads
to counterintuitive behavior that can confuse developers and introduce bugs.</p>
<h3>What is the potential impact?</h3>
<h4>Deferred execution in TaskGroup</h4>
<p>In asyncio’s TaskGroup, control flow statements don’t take immediate effect. Instead, they wait for all tasks in the group to complete before
executing. This can lead to:</p>
<ul>
  <li> Unexpected delays when tasks run longer than anticipated </li>
  <li> Code that appears to exit early but actually continues running </li>
  <li> Potential infinite loops if background tasks never complete </li>
</ul>
<h4>Lost return values in Nurseries</h4>
<p>In Trio and AnyIO nurseries, return values can be lost if other tasks in the nursery raise exceptions:</p>
<ul>
  <li> When a background task raises an exception, the return value from the main flow is discarded </li>
  <li> The nursery’s exception handling takes precedence over return values </li>
  <li> Silent data loss that’s difficult to debug </li>
</ul>
<h2>How to fix it in Asyncio</h2>
<p>Move the control flow statement outside the TaskGroup or Nursery block, and use the appropriate cancellation mechanism before exiting the
block.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import asyncio

async def process():
    async with asyncio.TaskGroup() as tg:
        tg.create_task(background_task())

        if condition():
            return "result"  # Noncompliant: waits for background_task() to complete
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import asyncio

async def process():
    result = None
    async with asyncio.TaskGroup() as tg:
        task = tg.create_task(background_task())

        if condition():
            result = "result"
            task.cancel()

    return result
</pre>
<h2>How to fix it in Trio</h2>
<p>Move the control flow statement outside the Nursery block, and use the appropriate cancellation mechanism before exiting the block.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
import trio

async def process():
    async with trio.open_nursery() as nursery:
        nursery.start_soon(background_task)

        if condition():
            return "result"  # Noncompliant: value may be lost
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
import trio

async def process():
    result = None
    async with trio.open_nursery() as nursery:
        nursery.start_soon(background_task)

        if condition():
            result = "result"
            nursery.cancel_scope.cancel()

    return result
</pre>
<h2>How to fix it in AnyIO</h2>
<p>Move the control flow statement outside the TaskGroup block, and use the appropriate cancellation mechanism before exiting the block.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="3" data-diff-type="noncompliant">
import anyio

async def process():
    async with anyio.create_task_group() as tg:
        tg.start_soon(background_task)

        if condition():
            return "result"  # Noncompliant: waits for background_task
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="3" data-diff-type="compliant">
import anyio

async def process():
    result = None
    async with anyio.create_task_group() as tg:
        tg.start_soon(background_task)

        if condition():
            result = "result"
            tg.cancel_scope.cancel()

    return result
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Asyncio documentation - <a href="https://docs.python.org/3/library/asyncio-task.html#task-groups">Task Groups</a> </li>
  <li> Trio documentation - <a href="https://trio.readthedocs.io/en/latest/reference-core.html#nurseries-and-spawning">Nurseries and spawning</a>
  </li>
  <li> AnyIO documentation - <a href="https://anyio.readthedocs.io/en/stable/tasks.html#creating-and-managing-tasks">Creating and managing tasks</a>
  </li>
  <li> Trio issue #1493 - <a href="https://github.com/python-trio/trio/issues/1493">Returning a value from inside a nursery block behaves
  counterintuitively</a> </li>
</ul>

