diff --git a/Python.g b/Python.g
index d3d4c2b..02cbb7a 100644
--- a/Python.g
+++ b/Python.g
@@ -121,6 +121,7 @@ small_stmt
         | global_stmt
         | nonlocal_stmt 
         | assert_stmt
+        | print_stmt
         ;
 
 del_stmt        : DEL exprlist;
@@ -156,6 +157,10 @@ import_as_names : import_as_name ( COMMA import_as_name )* COMMA?;
 dotted_as_names : dotted_as_name ( COMMA dotted_as_name )*;
 dotted_name     : IDENTIFIER ( DOT IDENTIFIER )*;
 
+print_stmt      : 'print'
+                ( testlist | RIGHTSHIFT testlist)?
+                ;
+
 
 //// $<Compound statements ////
 compound_stmt
@@ -351,7 +356,7 @@ argument    : test comp_for?
 STRINGLITERAL   : STRINGPREFIX? ( SHORTSTRING | LONGSTRING );
 
 fragment STRINGPREFIX
-        : ( 'r' | 'R' );
+        : ( 'r' | 'R'  | 'u' | 'U');
 
 fragment SHORTSTRING
         : '"' ( ESCAPESEQ | ~( '\\'|'\n'|'"' ) )* '"'
@@ -439,11 +444,11 @@ fragment NONZERODIGIT
 fragment DIGIT
         : '0' .. '9';
 fragment OCTINTEGER
-        : '0' ( 'o' | 'O' ) OCTDIGIT+;
+        : '0' ( 'o' | 'O' ) OCTDIGIT+ ( 'l' | 'L' )?;
 fragment HEXINTEGER
-        : '0' ( 'x' | 'X' ) HEXDIGIT+;
+        : '0' ( 'x' | 'X' ) HEXDIGIT+ ( 'l' | 'L' )?;
 fragment BININTEGER
-        : '0' ( 'b' | 'B' ) BINDIGIT+;
+        : '0' ( 'b' | 'B' ) BINDIGIT+ ( 'l' | 'L' )?;
 fragment OCTDIGIT
         : '0' .. '7';
 fragment HEXDIGIT
